### PR TITLE
Add gradients for variable losses

### PR DESCRIPTION
**Pull request by wielandbrendel**
So far gradients in Foolbox are always computed w.r.t. the cross-entropy loss. Many attacks, e.g. DeepFool and Carlini & Wagner, rely on other loss functions. This PR is supposed to fix this problem but is WIP. To Do:

- [x] Keras model
- [x] MXNet model
- [x] Theano model
- [x] Lasagne model
- [x] Tensorflow model

One breaking incompatibility is the fact that the default gradient being returned is now the gradient of the logits (I really don't see any reason to favour cross-entropy above anything else). In return some attacks have to be adapted accordingly (i.e. we need to add loss='crossentropy').

### CODE REVIEW

**Comment by jonasrauber** [comment]
Superseeded by supporting arbitrary losses using the `backward` method (backprop anything).
